### PR TITLE
[3.1] Pinned Builds manual dispatch workflow in CI

### PR DESCRIPTION
**Pull request by oschwaldp-oci**
Pinned builds run in GitHub Actions

Resolves: https://github.com/AntelopeIO/leap/issues/735

### CODE REVIEW

File: .github/workflows/pinned_build.yaml [hunk 1]
```diff
@@ -26,3 +26,45 @@ jobs:
...
+          ./scripts/pinned_build.sh deps build "$(nproc)"
+          cd build
+          cpack
+          ls -l
+          cd ..
+      - name: Upload package
+        uses: actions/upload-artifact@v3
+        with:
+          name: leap-${{matrix.platform}}-amd64
+          path: build/leap*.deb
```
Comment (spoonincode) [inline .github/workflows/pinned_build.yaml:66]: `leap-dev.deb` is broken for pinned builds, so I'd suggest only uploading `leap.deb` as part of this artifact somehow.

I'm also unsure whether or not we should have the artifact name explicitly call out pinned (e.g. `leap-${{matrix.platform}}-pinned-amd64`).. but I think it's okay like this for now

File: .github/workflows/pinned_build.yaml [hunk 2]
```diff
@@ -26,3 +26,45 @@ jobs:
           platform-file: .cicd/platforms.json
           password: ${{secrets.GITHUB_TOKEN}}
           package-name: builders
+
+  Build:
+    name: Build
+    needs: [d]
```
Comment (spoonincode) [inline .github/workflows/pinned_build.yaml:32]: we're not using the cached build environments in this job, so you can completely remove the "Discover Platforms" job

Comment (oschwaldp-oci) [reply .github/workflows/pinned_build.yaml:32]: Yeah, I put that in simply to not have a completely empty initial workflow file in https://github.com/AntelopeIO/leap/pull/1290. I'll remove it here.

Comment (oschwaldp-oci) [reply .github/workflows/pinned_build.yaml]: Addressed: [7c10660](https://github.com/AntelopeIO/leap/pull/1291/commits/7c106605cc4ac138c3604c9fc659913a9e28cb4c)

File: .github/workflows/pinned_build.yaml [hunk 3]
```diff
@@ -26,3 +26,45 @@ jobs:
...
+      - name: Install dependencies
+        run: |
+          # https://github.com/actions/runner/issues/2033
+          chown -R $(id -u):$(id -g) $PWD
+          ./scripts/install_deps.sh
+      - name: Build Pinned Build
+        run: |
+          ./scripts/pinned_build.sh deps build "$(nproc)"
+          cd build
+          cpack
```
[the remark this answers is not in the document]
Comment (oschwaldp-oci) [reply .github/workflows/pinned_build.yaml]: Addressed: [da6a068](https://github.com/AntelopeIO/leap/pull/1291/commits/da6a0686390f1ac773b7eee833c795aba343e118)

File: .github/workflows/pinned_build.yaml [hunk 4]
```diff
@@ -26,3 +26,45 @@ jobs:
...
+  Build:
+    name: Build
+    needs: [d]
+    if: always() && needs.d.result == 'success'
+    strategy:
+      fail-fast: false
+      matrix:
+        platform: [ubuntu18, ubuntu20, ubuntu22]
+    runs-on: ["self-hosted", "enf-x86-beefy"] # not sure if this warrants a different machine, but start here due to building
+    container: ${{ matrix.platform == 'ubuntu18' && 'ubuntu:bionic' || matrix.platform == 'ubuntu20' && 'ubuntu:focal' || 'ubuntu:jammy' }}
```
[the remark this answers is not in the document]
Comment (oschwaldp-oci) [reply .github/workflows/pinned_build.yaml:23]: Looks like this is working from initial testing of the workflow, I'm seeing `bionic`, `focal`, and `jammy`.

File: .github/workflows/pinned_build.yaml [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+          chown -R $(id -u):$(id -g) $PWD
+          ./scripts/install_deps.sh
+      - name: Build Pinned Build
+        run: |
+          ./scripts/pinned_build.sh deps build "$(nproc)"
+      - name: Check and select artifact
+        run: |
+          cd build
+          echo "PINNED_BUILD_ARTIFACT="$(ls | grep -E 'leap-[^d]+*.deb')"" >> "$GITHUB_ENV"
+          echo "PINNED_BUILD_NAME="$(ls | grep -E 'leap-[^d]+*.deb' | sed -e 's/-x86_64.deb/-pinned-amd64/')"" >> "$GITHUB_ENV"
```
Comment (spoonincode) [inline .github/workflows/pinned_build.yaml:53]: What's the need for these shenanigans? I know the 3.1's .deb files are named a little differently which makes the difference between `leap` & `leap-dev` trickier. What stops us from something like this for 3.1,
```yaml
      - name: Upload package
        uses: actions/upload-artifact@v3
        with:
          name: leap-${{matrix.platform}}-pinned-amd64
          path: build/leap-3*.deb
```
and then we can make it more generic on 3.2+ with
```yaml
      - name: Upload package
        uses: actions/upload-artifact@v3
        with:
          name: leap-${{matrix.platform}}-pinned-amd64
          path: build/leap_*.deb
```

Comment (oschwaldp-oci) [reply .github/workflows/pinned_build.yaml:53]: Those shenanigans allow for selecting the `.deb` file a little more generically without having to specify the leading major version number (The wildcard matching for `upload-artifact` didn't seem to support full regex) and then crafting upload names that matched more closely what I saw in the artifacts on our releases which would include the version numbers, but then add in the `pinned` identifier:

Release assets are named as such:
image

The pinned build upload assets are named:
image

Comment (spoonincode) [reply .github/workflows/pinned_build.yaml:53]: I don't think it's necessary for the artifact name to have the version number in it, and quite possibly we _shouldn't_ do that. It makes things harder when we want to make use of a build that isn't a release; a perfect example of that is in the libtester workflow where you simply grab the `cdt.deb` from main: what version number would you expect to find in that situation? It's difficult to know ahead of time, and then coupled with `asset-artifact-download-action` not supporting wildcards for the `artifact-name` name currently, it seems to break the use case of another repo's workflow being able to find and use any pinned build.

Admittedly once we do #1101 we're going to have to iterate over the artifacts in a workflow and look for some pattern.. but that pattern likely wouldn't need the version numbers still.

> selecting the .deb file a little more generically without having to specify the leading major version number (The wildcard matching for upload-artifact didn't seem to support full regex

afaik this is strictly a 3.1 problem, and shouldn't be a problem 3.2+, so I don't see any real problem with hardcoding the "3" in the 3.1 branch's workflow. Would the yaml samples in my above comment work?

Comment (oschwaldp-oci) [reply .github/workflows/pinned_build.yaml:53]: Sure. That's fine. I can do it that way and follow that naming convention if you'd prefer.  Would you like the artifact to have the `.deb` file extension in the name?

**Review by spoonincode — CHANGES_REQUESTED** [review]

File: .github/workflows/pinned_build.yaml [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+          ./scripts/pinned_build.sh deps build "$(nproc)"
+      - name: Upload package
+        uses: actions/upload-artifact@v3
+        with:
+          name: leap-${{matrix.platform}}-pinned-amd64
+          path: build/leap-3*.deb
+      - name: Run Parallel Tests
+        run: |
+          cd build
+          ctest --output-on-failure -j $(nproc) -LE "(nonparallelizable_tests|long_running_tests)" --timeout 420
```
[the remark this answers is not in the document]
Comment (spoonincode) [reply .github/workflows/pinned_build.yaml:56]: oookay since it just seems to be flaky let's go

File: .github/workflows/pinned_build.yaml [hunk 7]
```diff
@@ -0,0 +1,56 @@
+name: "Pinned Build"
+
+on:
+  workflow_dispatch:
+
+permissions:
+  packages: read
```
Comment (spoonincode) [inline .github/workflows/pinned_build.yaml:7]: ah, we could remove this permission I think. nbd tho

File: .github/workflows/pinned_build.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (oschwaldp-oci) [reply .github/workflows/pinned_build.yaml:56]: Manual Pinned Build run example here: https://github.com/AntelopeIO/leap/actions/runs/5346832312/jobs/9695391713

File: .github/workflows/pinned_build.yaml (hunk 5 above)
Comment (oschwaldp-oci) [reply .github/workflows/pinned_build.yaml]: image

**Pull request merged**